histoqc/SaveModule.py

```diff
@@ -19,6 +20,96 @@ def blend2Images(img, mask):
     return out
 
 
+'''
+the followings are helper functions for generating geojson
+'''
+
+feature_template = {
+    "type": "Feature",
+    "id": None,
+    "geometry": {
+        "type": None,
+        "coordinates": []
+    },
+    "properties": {
+        "objectType": "annotation"
+    }
+}
+
+feature_collection_template = {
+  "type": "FeatureCollection",
+  "features": []
+}
+
+
+
+
+
+
+
+def binaryMask2Geojson(s, mask):
+    # get the dimension of slide
+    (dim_width, dim_height) = s['os_handle'].dimensions
+    # get the dimension of mask
+    (mask_height, mask_width) = mask.shape
+
+    # convert binary mask to contours
+    # contours = measure.find_contours(mask)
+    contours, hierarchy = cv2.findContours(img_as_ubyte(mask), cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+    children = [[] for i in range(len(contours))]
+    for i, cnt in enumerate(contours):
+        # first_child_idx = hier[0, i, 2]
+        parent_idx = hierarchy[0, i, 3]
+        if (parent_idx == -1):
+            continue
+        # add contour to parent's children node
+        children[parent_idx].append(cnt)    
+    
+    # no contours detected
+    if not len(contours):
+        # set default format
+        ann_format = "xml"
+        # warning msg
+        msg = f"No contour detected at use mask image. Geojson annotation won't be generated."
+        logging.warning(f"{s['filename']} - {msg}")
+        s["warnings"].append(msg)
+        return None
+    
+    # copy feature collection template in geojson
+    feature_collection = deepcopy(feature_collection_template)
+    for i, contour in enumerate(contours):
+        first_child_idx = hierarchy[0, i, 2]
+        parent_idx = hierarchy[0, i, 3]
+        
+        if (parent_idx != -1):
+            continue
+        
+        # copy feature template in geojson
+        new_feature = deepcopy(feature_template)
+        # set id
+        new_feature["id"] = uuid.uuid4().hex
+        # scale up the coordinate
+        # points = np.asarray(np.flip(contour / [mask_height, mask_width] * [dim_height, dim_width]),dtype="int")
+        points = np.asarray(contour / [mask_height, mask_width] * [dim_height, dim_width],dtype="int")
+        points = np.append(points, [points[0]], axis=0)
+        points = points[:,0]
+
+        if first_child_idx == -1:
+            new_feature['geometry']['type'] = 'Polygon'
+            new_feature['geometry']['coordinates'].append(points.tolist())       
+        else:
+            new_feature['geometry']['type'] = 'MultiPolygon'
```

Comment:
I discovered last week that MultiPolygon is not actually the appropriate type for representing polygons with holes. Instead, the Polygon data structure should be used, e.g., 
```json
{
  "type": "Feature",
  "properties": {},
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          -120,
          60
        ],
        [
          120,
          60
        ],
        [
          120,
          -60
        ],
        [
          -120,
          -60
        ],
        [
          -120,
          60
        ]
      ],
      [
        [
          -60,
          30
        ],
        [
          60,
          30
        ],
        [
          60,
          -30
        ],
        [
          -60,
          -30
        ],
        [
          -60,
          30
        ]
      ]
    ]
  }
}
```
... where the first set of coordinates represents the outer ring and the second represents the inner hole.
https://gist.github.com/andrewharvey/971b9db1900a62efa7635f6a6d337ae7#file-polygon-with-a-hole-geojson

Also see this GeoJS sandbox. If you draw a polygon with a hole it will be stored as a "Polygon" type.

Sorry for the confusion. Can you change the code to only use the Polygon type?

Reply:
Yes. We briefly discussed something relevant in #258 

But I think multipolygon support should be retained as it is quite common for softwares (e.g., PathML) to have cascaded annotation support - which groups a list of disconnected polygons in a confined region together and depending on user's choice and their use cases an optional multipolygon output can be helpful.

Existing code in annotation module could be a good example for illustration (though it's geojson to mask not the other way around): https://github.com/choosehappy/HistoQC/blob/master/histoqc/annotations/annotation/geojson.py#25

Reply:
Thanks Yufei.

The MultiPolygon support that you are describing would still require a change to the code, to produce e.g.,

```
{
  "type": "Feature",
  "properties": {},
  "geometry": {
    "type": "MultiPolygon",
    "coordinates": [[poly1_ring_coords_array, poly1_hole_coords_array], [poly2_ring_coords_array, poly2_hole_coords_array]]
   }
}
```

Correct me if I'm wrong, but it looks like @nanli-emory 's code produces a multipolygon like so:
```
{
  "type": "Feature",
  "properties": {},
  "geometry": {
    "type": "MultiPolygon",
    "coordinates": [[poly1_ring_coords_array], [poly1_hole_coords_array]]
   }
}
```
... where poly1_ring_coords_array and poly1_hole_coords_array are not actually sibling arrays in the nested structure.

Reply:
Hi @jacksonjacobs1 , I use geojson lib to generate FeatureCollection. Please review it. Thanks,